cloud/run/workflow.py

```diff
@@ -0,0 +1,182 @@
+    macro_processor: Optional[MacroProcessor] = None,
+    executor_factory: Type[Executor] = SynchronousExecutor,
+) -> None:
+    """Executes the preprocess->translate->postprocess workflow.
+
+    Example:
+        .. code-block::
+
+            from run.gcp.bqms.request import build as build_bqms_request
+            from run.hooks import (
+                postprocess as postprocess_hook,
+                preprocess as preprocess_hook
+            )
+            from run.macro_processor import MacroProcessor
+            from run.paths import Paths
+
+            paths = Paths.from_mapping(...)
+            bqms_request = build_bqms_request(...)
+            macro_processor = MacroProcessor.from_mapping(...)
+            execute(paths, preprocess_hook, postprocess_hook, bqms_request,
+                    macro_processor)
+
+    Args:
+        paths: A run.paths.Paths instance containing project paths.
+        preprocess_hook: A callable for hooking user-defined preprocessing
+            logic into the translation workflow.
+        postprocess_hook: A callable for hooking user-defined postprocessing
+            logic into the translation workflow.
+        bqms_request: A
+            google.cloud.bigquery_migration_v2.CreateMigrationWorkflowRequest
+            for requesting the execution of the BQMS migration workflow.
+        macro_processor: An optional run.macro_processor.MacroProcessor for
+            handling macro/templating languages embedded in code to be
+            translated.
+        executor_factory: An optional concurrent.futures.Executor factory used
+            to create an executor for per-path pre and postprocessing tasks.
+    """
+
+    def _preprocess(source_file_path: Path) -> None:
+        """Preprocessing task.
+
+        To be submitted to executor for each input path.
+        """
+        relative_file_path = source_file_path.relative_to(paths.input_path)
+        target_file_path = paths.preprocessed_path / relative_file_path
+        target_file_path.parent.mkdir(parents=True, exist_ok=True)
+
+        # Do not preprocess metadata zip file.
+        if relative_file_path.name.endswith(".zip"):
+            with source_file_path.open(mode="rb") as source_file:
+                source_bytes = source_file.read()
+            with target_file_path.open(mode="wb") as target_file:
+                target_file.write(source_bytes)
+            return
+
+        with source_file_path.open(mode="r", encoding="utf-8") as source_file:
+            source_text = source_file.read()
+
+        preprocessed_text = preprocess_hook(relative_file_path, source_text)
+        macro_expanded_text = (
+            macro_processor.expand(relative_file_path, preprocessed_text)
+            if macro_processor
+            else preprocessed_text
+        )
+
+        with target_file_path.open(mode="w", encoding="utf-8") as target_file:
+            target_file.write(macro_expanded_text)
```

Comment:
So it seems cloud run can use gcs for file IO directly?  This is pretty cool.

Reply:
Actually, I am using [cloudpathlib](https://cloudpathlib.drivendata.org/stable/) which abstracts over local paths and cloud paths like GCS paths. It's a very handy small library with just a few classes. In fact, if we installed the object store libs from AWS and Azure (right now we only install the GCS lib), we could support reading and writing raw input and postprocessed output to those clouds as well.